File: SRC/dormhr.f
      SUBROUTINE DORMHR( SIDE, TRANS, M, N, ILO, IHI, A, LDA, TAU, C, LDC, WORK, LWORK, INFO )

*  -- LAPACK computational routine --
*  -- LAPACK is a software package provided by Univ. of Tennessee,    --
*  -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--

      // .. Scalar Arguments ..
      String             SIDE, TRANS;
      int                IHI, ILO, INFO, LDA, LDC, LWORK, M, N;
      // ..
      // .. Array Arguments ..
      double             A( LDA, * ), C( LDC, * ), TAU( * ), WORK( * );
      // ..

*  =====================================================================

      // .. Local Scalars ..
      bool               LEFT, LQUERY;
      int                I1, I2, IINFO, LWKOPT, MI, NB, NH, NI, NQ, NW;
      // ..
      // .. External Functions ..
      bool               LSAME;
      int                ILAENV;
      // EXTERNAL LSAME, ILAENV
      // ..
      // .. External Subroutines ..
      // EXTERNAL DORMQR, XERBLA
      // ..
      // .. Intrinsic Functions ..
      // INTRINSIC MAX, MIN
      // ..
      // .. Executable Statements ..

      // Test the input arguments

      INFO = 0
      NH = IHI - ILO
      LEFT = LSAME( SIDE, 'L' )
      LQUERY = ( LWORK == -1 )

      // NQ is the order of Q and NW is the minimum dimension of WORK

      if ( LEFT ) {
         NQ = M
         NW = MAX( 1, N )
      } else {
         NQ = N
         NW = MAX( 1, M )
      }
      if ( .NOT.LEFT && .NOT.LSAME( SIDE, 'R' ) ) {
         INFO = -1
      } else if ( .NOT.LSAME( TRANS, 'N' ) && .NOT.LSAME( TRANS, 'T' ) ) {
         INFO = -2
      } else if ( M < 0 ) {
         INFO = -3
      } else if ( N < 0 ) {
         INFO = -4
      } else if ( ILO < 1 || ILO > MAX( 1, NQ ) ) {
         INFO = -5
      } else if ( IHI < MIN( ILO, NQ ) || IHI > NQ ) {
         INFO = -6
      } else if ( LDA < MAX( 1, NQ ) ) {
         INFO = -8
      } else if ( LDC < MAX( 1, M ) ) {
         INFO = -11
      } else if ( LWORK < NW && .NOT.LQUERY ) {
         INFO = -13
      }

      if ( INFO == 0 ) {
         if ( LEFT ) {
            NB = ILAENV( 1, 'DORMQR', SIDE // TRANS, NH, N, NH, -1 )
         } else {
            NB = ILAENV( 1, 'DORMQR', SIDE // TRANS, M, NH, NH, -1 )
         }
         LWKOPT = NW*NB
         WORK( 1 ) = LWKOPT
      }

      if ( INFO != 0 ) {
         xerbla('DORMHR', -INFO );
         RETURN
      } else if ( LQUERY ) {
         RETURN
      }

      // Quick return if possible

      if ( M == 0 || N == 0 || NH == 0 ) {
         WORK( 1 ) = 1
         RETURN
      }

      if ( LEFT ) {
         MI = NH
         NI = N
         I1 = ILO + 1
         I2 = 1
      } else {
         MI = M
         NI = NH
         I1 = 1
         I2 = ILO + 1
      }

      dormqr(SIDE, TRANS, MI, NI, NH, A( ILO+1, ILO ), LDA, TAU( ILO ), C( I1, I2 ), LDC, WORK, LWORK, IINFO );

      WORK( 1 ) = LWKOPT
      RETURN

      // End of DORMHR

      }

File: lib/src/dormhr.dart
// Copyright (c) 2024 Guilherme Lepsch. All rights reserved. Use of this
// source code is governed by a BSD-style license that can be found in the
// [LICENSE file](https://github.com/lepsch/dart_lapack/blob/main/LICENSE).

import 'dart:math';

import 'package:dart_lapack/src/install/lsame.dart';
import 'package:dart_lapack/src/box.dart';
import 'package:dart_lapack/src/dormqr.dart';
import 'package:dart_lapack/src/ilaenv.dart';
import 'package:dart_lapack/src/matrix.dart';
import 'package:dart_lapack/src/xerbla.dart';

void dormhr(
  final String SIDE,
  final String TRANS,
  final int M,
  final int N,
  final int ILO,
  final int IHI,
  final Matrix<double> A_,
  final int LDA,
  final Array<double> TAU_,
  final Matrix<double> C_,
  final int LDC,
  final Array<double> WORK_,
  final int LWORK,
  final Box<int> INFO,
) {
// -- LAPACK computational routine --
// -- LAPACK is a software package provided by Univ. of Tennessee,    --
// -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--
  final A = A_.having(ld: LDA);
  final TAU = TAU_.having();
  final C = C_.having(ld: LDC);
  final WORK = WORK_.having();
  bool LEFT, LQUERY;
  int I1, I2, LWKOPT = 0, MI, NB, NH, NI, NQ, NW;
  final IINFO = Box(0);

  // Test the input arguments

  INFO.value = 0;
  NH = IHI - ILO;
  LEFT = lsame(SIDE, 'L');
  LQUERY = (LWORK == -1);

  // NQ is the order of Q and NW is the minimum dimension of WORK

  if (LEFT) {
    NQ = M;
    NW = max(1, N);
  } else {
    NQ = N;
    NW = max(1, M);
  }
  if (!LEFT && !lsame(SIDE, 'R')) {
    INFO.value = -1;
  } else if (!lsame(TRANS, 'N') && !lsame(TRANS, 'T')) {
    INFO.value = -2;
  } else if (M < 0) {
    INFO.value = -3;
  } else if (N < 0) {
    INFO.value = -4;
  } else if (ILO < 1 || ILO > max(1, NQ)) {
    INFO.value = -5;
  } else if (IHI < min(ILO, NQ) || IHI > NQ) {
    INFO.value = -6;
  } else if (LDA < max(1, NQ)) {
    INFO.value = -8;
  } else if (LDC < max(1, M)) {
    INFO.value = -11;
  } else if (LWORK < NW && !LQUERY) {
    INFO.value = -13;
  }

  if (INFO.value == 0) {
    if (LEFT) {
      NB = ilaenv(1, 'DORMQR', SIDE + TRANS, NH, N, NH, -1);
    } else {
      NB = ilaenv(1, 'DORMQR', SIDE + TRANS, M, NH, NH, -1);
    }
    LWKOPT = NW * NB;
    WORK[1] = LWKOPT.toDouble();
  }

  if (INFO.value != 0) {
    xerbla('DORMHR', -INFO.value);
    return;
  } else if (LQUERY) {
    return;
  }

  // Quick return if possible

  if (M == 0 || N == 0 || NH == 0) {
    WORK[1] = 1;
    return;
  }

  if (LEFT) {
    MI = NH;
    NI = N;
    I1 = ILO + 1;
    I2 = 1;
  } else {
    MI = M;
    NI = NH;
    I1 = 1;
    I2 = ILO + 1;
  }

  dormqr(SIDE, TRANS, MI, NI, NH, A(ILO + 1, ILO), LDA, TAU(ILO), C(I1, I2),
      LDC, WORK, LWORK, IINFO);

  WORK[1] = LWKOPT.toDouble();
}
